### ISSUE TITLE
Macro define a class name to a different one causes extra whitespace to appear

### ISSUE THREAD

**Issue by scott-zhong** [doxygen/doxygen#10762]
**Describe the bug**
Macro define a class name to a different one causes extra whitespace to appear when `MACRO_EXPANSION = YES` is used. The problem started with 1.8.12 and continues with Doxygen trunk (https://github.com/doxygen/doxygen/commit/cbb48edb43715e3420a69d69c21e6e3c4f998894).

```
#define A B

/**
 * A docs
 */
class A
{
public:
    /**
     * Type of food.
     */
    enum Food {
        /**
         * apple
         */
        apple  
    };
};

/**
 * Test docs
 */
template <class T>
struct Test
{
};

/**
 * Test specialized docs
 */
template<>
struct Test<A>
{
    /**
     * Test specialized ctor
     */
    Test(A::Food food)
        : food_(food) {}

    A::Food food_;
};
```

There is an extra whitespace going from Doxygen 1.8.11 to 1.8.12 and later (`B::Food` -> `B ::Food`).

**Screenshots**
Doxygen 1.8.11:
![image](https://github.com/doxygen/doxygen/assets/32584126/fef26cce-0e2e-4fc8-8ec0-32cd095c252e)

Doxygen 1.8.12 - Doxygen trunk (https://github.com/doxygen/doxygen/commit/cbb48edb43715e3420a69d69c21e6e3c4f998894):
![image](https://github.com/doxygen/doxygen/assets/32584126/f0b0a598-8a65-46bd-a7f9-10ffde544130)

**To Reproduce**
[testcase.zip](https://github.com/doxygen/doxygen/files/14676036/testcase.zip)

**Expected behavior**
No extra whitespace in the generated documentation.

**Comment by scott-zhong** [comment]
The whitespace also appears with templates as well.

```
/**
 * Base docs
 */
template <class Type>
class Base
{
public:
    /**
     * A typedef
     */
    typedef typename Type::element_type element_type;
};

/**
 * MyCharType docs
 */
class MyCharType
{
public:
    /**
     * A typedef
     */
    typedef char element_type;
};

/**
 * Derived docs
 */
class Derived : public Base<MyCharType>
{
};
```

Doxygen 1.8.11:
![image](https://github.com/doxygen/doxygen/assets/32584126/128e59bf-d628-436c-95f4-cff9f9d07498)

Doxygen 1.8.12:
![image](https://github.com/doxygen/doxygen/assets/32584126/12451ab6-7f50-452b-9f33-934953b9b005)

The problem is the extra whitespace `MyCharType::element_type` -> `MyCharType ::element_type`.  I don't know if trunk will have the same problem because of the bug https://github.com/doxygen/doxygen/issues/10760 .

**Comment by albert-github** [comment]
A bisect showed:
```
00ee930a1d73e11885197102c54fd4c8141127da is the first bad commit
commit 00ee930a1d73e11885197102c54fd4c8141127da
Date:   Sun Mar 13 10:12:46 2016 +0100

    reimplemented removeRedundantWhiteSpace() to improve performance

 src/util.cpp | 385 +++++++++++++++++++++++++++++++++++------------------------
 1 file changed, 231 insertions(+), 154 deletions(-)
```
Looks like a bit of a side effect.

I've just pushed a proposed patch, pull request #10771

**Comment by albert-github** [comment]
An adjusted implementation has been integrated in master on GitHub

See https://github.com/doxygen/doxygen/pull/10771#issuecomment-2029324943 for the remark about this adjusted implementation.

**Comment by doxygen** [comment]
See [this commit](https://github.com/doxygen/doxygen/commit/d5bc73a572688b6f2518b735602aeabc6915666d) for a better solution that doesn't remove spaces where they are needed.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.11.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
